### PR TITLE
interfaces/builtin: support time and date settings via 'org.freedesktop.timedate1

### PR DESCRIPTION
**Pull request by stolowski**
Added new timedate-control interface and updated two existing interfaces: timeserver-control and timezone-control to allow access to respective parts of systemd' org.freedesktop.timedate1 interface (lp #1616052). 

The test shamelessly stolen from unity7_test.go.

Tested introspection and a property getter manually with:

$ timedatetest.gdbus introspect --system --dest org.freedesktop.timedate1 --object-path /org/freedesktop/timedate1
$ timedatetest.gdbus call --system --dest org.freedesktop.timedate1 --object-path /org/freedesktop/timedate1 --method org.freedesktop.DBus.Properties.Get "org.freedesktop.timedate1" "Timezone"

with the following simple snap:

name: timedatetest
version: 1.0
summary: A test
description: A test

apps:
  gdbus:
    command: usr/bin/gdbus
    plugs: [timedate-control]

parts:
  gdbus:
    plugin: nil
    stage-packages:
        - libglib2.0-bin

### CODE REVIEW

File: interfaces/builtin/timedate_control.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    peer=(label=unconfined),
+
+dbus (send)
+    bus=system
+    path=/org/freedesktop/timedate1
+    interface=org.freedesktop.timedate1
+    member="Set{Time,Timezone,LocalRTC,NTP}"
+    peer=(label=unconfined),
+
+dbus (receive, send)
```
Comment (zyga) [inline interfaces/builtin/timedate_control.go]: Why send and receive? // my knowledge of apparmor is rusty here.

File: interfaces/builtin/timedate_control.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+send
+sendto
+sendmsg
+socket
+`
+
+// NewTimeDateControlInterface returns a new "time-date-control" interface.
+func NewTimeDateControlInterface() interfaces.Interface {
+	return &commonInterface{
+		name: "time-date-control",
```
Comment (zyga) [inline interfaces/builtin/timedate_control.go]: This is inconsistent with the file name, either call the module `time_data_control.go` or rename the interface to `timedate-control`.

**Comment by zyga** [comment]
Please document this interface so that we know what it allows to do (in `docs/interfaces.md`) and decide if it is reserved (restricted to trusted snaps) or not. I'd say it is and it should auto-connect.

I have one question about apparmor, please ask @jdstrand for a review after you know the answer to my question.

File: interfaces/builtin/timedate_control.go (hunk 2 above)
Comment (stolowski) [inline interfaces/builtin/timedate_control.go]: Ok, renamed to timedate-control.

File: interfaces/builtin/timedate_control.go (hunk 1 above)
Comment (stolowski) [inline interfaces/builtin/timedate_control.go]: Fair point. receive is needed for property change signals (and only for them) and it was a little bit too lenient. I've split this block into two and added comments - property change signals need 'receive', Get & GetAll need 'send'.

**Comment by stolowski** [comment]
@zyga I've added a short doc and made it auto-connect. I'm not sure however about the making it 'reserved', is it about marking it reservedForOs (which it is already)?

**Comment by niemeyer** [comment]
Needs conflict solving, and a review from @jdstrand. LGTM other than that.

File: interfaces/builtin/timedate_control.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+`
+
+// NewTimeDateControlInterface returns a new "timedate-control" interface.
+func NewTimeDateControlInterface() interfaces.Interface {
+	return &commonInterface{
+		name: "timedate-control",
+		connectedPlugAppArmor: timeDateControlConnectedPlugAppArmor,
+		connectedPlugSecComp:  timeDateControlConnectedPlugSecComp,
+		reservedForOS:         true,
+		autoConnect:           true,
```
Comment (jdstrand) [inline interfaces/builtin/timedate_control.go]: Adjusting the time on a system is a privileged action and this should not autoconnect.

File: interfaces/builtin/timedate_control.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    member=Get{,All}
+    peer=(label=unconfined),
+
+# Receive property changed events
+dbus (receive)
+    bus=system
+    path=/org/freedesktop/timedate1
+    interface=org.freedesktop.DBus.Properties
+    member=PropertiesChanged
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/timedate_control.go]: Unfortunately we can't mediate on method data and systemd only provides Get methods for properties, so you'll need to have these two rules in timezone-control and timeserver-control. This means that all the time* interfaces will be able to read each others properties, but at least they won't be able to Set them.

File: interfaces/builtin/timedate_control.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    path=/org/freedesktop/timedate1
+    interface=org.freedesktop.DBus.Introspectable
+    member=Introspect
+    peer=(label=unconfined),
+
+dbus (send)
+    bus=system
+    path=/org/freedesktop/timedate1
+    interface=org.freedesktop.timedate1
+    member="Set{Time,Timezone,LocalRTC,NTP}"
```
Comment (jdstrand) [inline interfaces/builtin/timedate_control.go]: We already have the timeserver-control interface for NTP. NTP calls should be moved there.

We already have the timezone-control interface for adjusting TZ. TZ calls should be moved there.

SetTime and SetTimeZone can remain here.

**Comment by jdstrand** [comment]
'time-control' is probably a better name than timedate-control. Will want @niemeyer to weigh in on that.

**Comment by stolowski** [comment]
@jdstrand Thanks for the review, I have followed your suggestions and moved some permissions to the other interfaces. All interfaces have permissions to read the properties as well as to introspect timedate1 object.

File: interfaces/builtin/timedate_control_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+func (s *TimeDateControlTestInterfaceSuite) TestSanitizeIncorrectInterface(c *C) {
+	c.Assert(func() { s.iface.SanitizeSlot(&interfaces.Slot{SlotInfo: &snap.SlotInfo{Interface: "other"}}) },
+		PanicMatches, `slot is not of interface "timedate-control"`)
+	c.Assert(func() { s.iface.SanitizePlug(&interfaces.Plug{PlugInfo: &snap.PlugInfo{Interface: "other"}}) },
+		PanicMatches, `plug is not of interface "timedate-control"`)
+}
+
+func (s *TimeDateControlTestInterfaceSuite) TestUnusedSecuritySystems(c *C) {
```
Comment (zyga) [inline interfaces/builtin/timedate_control_test.go]: Let's drop this test please. We just concluded a large simplification where those tests were axed.

Comment (stolowski) [inline interfaces/builtin/timedate_control_test.go]: Ok, removed.

File: docs/interfaces.md [hunk 7]
```diff
@@ -463,6 +463,13 @@ trusted apps.
...
+### timedate-control
+
+Can modify time/date exposed via D-Bus by systemd-timedated and read all
+all properties of /org/freedesktop/timedate1 D-Bus object.
+
+* Auto-Connect: yes
+
 ### timeserver-control
 
 Can manage timeservers directly separate from ``config core``.
```
Comment (jdstrand) [inline docs/interfaces.md]: Can you update this to be:
```
Can manage timeservers via systemd-timedated and directly separate from ``config core``

### timezone-control

Can manage timezone via systemd-timedated and directly separate from ``config core``

* Auto-Connect: no
```

In other words, update timeserver-control to reference systemd-timedated and add timezone-control.

File: snap/implicit.go [hunk 8]
```diff
@@ -54,6 +54,7 @@ var implicitSlots = []string{
 	"timezone-control",
 	"tpm",
 	"kernel-module-control",
+	"timedate-control",
```
Comment (jdstrand) [inline snap/implicit.go]: Can you put this in alphabetical order?

**Comment by jdstrand** [comment]
Note that I approved the changes you made, but still request further changes (using github's 'comment' mechanism instead of 'request changes' to avoid blocking on me).

**Comment by niemeyer** [comment]
+1 on `time-control` if we intend this interface to remain intermediating the ability to set the time even if we change the backend.

**Comment by jdstrand** [comment]
"+1 on time-control if we intend this interface to remain intermediating the ability to set the time even if we change the backend."

Yes this is the intent. In this manner it is similar to timezone-control and timeserver-control.

@stolowski - we now have agreement. Please rename to 'time-control' and I can do the final review.

**Comment by stolowski** [comment]
@jdstrand @niemeyer Ok, renamed to time-control and address other comments re docs (and I hope I didn't miss anything).

**Comment by jdstrand** [comment]
FYI, the request to add `timezone-control` to docs/interfaces.md was missed in this merge.

File: docs/interfaces.md (hunk 7 above)
Comment (stolowski) [inline docs/interfaces.md]: @jdstrand apologies for forgetting about this change, I'll address it in a separete PR.

**Pull request merged**